core/src/main/java/tc/oc/pgm/modules/StatsMatchModule.java

```diff
@@ -0,0 +1,299 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import net.md_5.bungee.api.ChatColor;
+import org.bukkit.Bukkit;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.Listener;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchModule;
+import tc.oc.pgm.api.match.MatchScope;
+import tc.oc.pgm.api.match.event.MatchFinishEvent;
+import tc.oc.pgm.api.match.factory.MatchModuleFactory;
+import tc.oc.pgm.api.module.exception.ModuleLoadException;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.api.player.PlayerRelation;
+import tc.oc.pgm.api.player.event.MatchPlayerDeathEvent;
+import tc.oc.pgm.events.ListenerScope;
+import tc.oc.pgm.tracker.damage.ProjectileInfo;
+import tc.oc.util.bukkit.component.Component;
+import tc.oc.util.bukkit.component.ComponentUtils;
+import tc.oc.util.bukkit.component.Components;
+import tc.oc.util.bukkit.component.types.PersonalizedText;
+import tc.oc.util.bukkit.component.types.PersonalizedTranslatable;
+import tc.oc.util.bukkit.nms.NMSHacks;
+import tc.oc.util.bukkit.translations.AllTranslations;
+
+@ListenerScope(MatchScope.RUNNING)
+public class StatsMatchModule implements MatchModule, Listener {
+
+  public static class Factory implements MatchModuleFactory<StatsMatchModule> {
+
+    @Override
+    public StatsMatchModule createMatchModule(Match match) throws ModuleLoadException {
+      return new StatsMatchModule();
+    }
+  }
+
+  public static class PlayerStats {
+    private int kills;
+    private int deaths;
+    private int killstreak;
+    private int killstreakMax;
+    private int longestBowKill;
+
+    private void onMurder() {
+      kills++;
+      killstreak++;
+      if (killstreak > killstreakMax) killstreakMax = killstreak;
+    }
+
+    private void onDeath() {
+      deaths++;
+      killstreak = 0;
+    }
+
+    private DecimalFormat decimalFormatBowshot = new DecimalFormat("#");
+
+    private void setLongestBowKill(double distance) {
+      if (distance > longestBowKill) {
+        longestBowKill = Integer.parseInt(decimalFormatBowshot.format(Math.round(distance)));
+      }
+    }
+
+    private DecimalFormat decimalFormatKd = new DecimalFormat("#.##");
+
+    Component getBasicStatsMessage() {
+      String kd;
+      if (deaths == 0) {
+        kd = "0";
+      } else {
+        kd = decimalFormatKd.format(kills / deaths);
+      }
+      return new Component(
+          new PersonalizedTranslatable(
+                  "stats.basic",
+                  new PersonalizedText(Integer.toString(kills), ChatColor.GREEN),
+                  new PersonalizedText(Integer.toString(killstreak), ChatColor.GREEN),
+                  new PersonalizedText(Integer.toString(deaths), ChatColor.RED),
+                  new PersonalizedText(kd, ChatColor.GREEN))
+              .render());
+    }
+  }
+
+  private static Map<UUID, PlayerStats> allPlayerStats;
+
+  public StatsMatchModule() {
+    allPlayerStats = new HashMap<>();
```

Comment:
This is defining static context from the instance context.

Reply:
What does that mean? haha

Reply:
`allPlayerStats` is a static variable. This means that the variable is global and not object depended. If you create a new instance of `StatsMatchModule` the contents of `allPlayerStats` should still be the same. The problem is that you are modifying static (global) variable from the object (its constructor). You should either make `allPlayerStats` non-static and make it object depended, or initialize it just once and make it a constant. If so, there will only be one instance of this `allPlayerStats` map.

```java
private final Map<UUID, PlayerStats> allPlayerStats = new HashMap<>();
```

or

```java
private static final Map<UUID, PlayerStats> ALL_PLAYER_STATS = new HashMap<>();
```

Reply:
What do i even need the constructor for? I only added it because Electroids example has it. The Module does not need to get passed any objects on creation anyways.., i will make i private static final,